### PR TITLE
refactor(api): move protocol impl classes to protocol_api.core

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This PR moves the existing Python Protocol API implementation interfaces and concrete implementations to a new, private module: `opentrons.protocol_api._core`. This refactor is primarily focused on lowering making it easier to add a new, ProtocolEngine-based PAPIv2 implementation.

This PR closes RCORE-79

## Changelog

- refactor(api): move protocol impl classes to protocol_api._core

## Review requests

- [ ] Does this change make sense / make this code easier to follow?
- [ ] Smoke test PAPIv2 analysis and execution via HTTP API
- [ ] Smoke test PAPIv2 analysis via `opentrons_simulate`
- [ ] Smoke test PAPIv2 execution via `opentrons_execute`

### Follow up work

1. Add skeleton `ProtocolEngine`-based cores behind a feature flag (RCORE-78)
2. Test drive through the public PAPI interfaces to reshape cores as needed
    - Probably rename some classes from `...Implementation` to `...Core`
    - Ensure that interfaces can function as facades for both the legacy direct-hardware execution logic and the newer, engine-based logic
    - Tickets to be written in RCORE-77 

## Risk assessment

Low, code shuffling only. PR increased type checking coverage of these important PAPIv2 interfaces

### CODE REVIEW

File: api/src/opentrons/execute.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 from opentrons.config import IS_ROBOT, JUPYTER_NOTEBOOK_LABWARE_DIR
 from opentrons.protocol_api import MAX_SUPPORTED_VERSION
 from opentrons.protocols.execution import execute as execute_apiv2
-from opentrons.protocols.context.protocol_api.protocol_context import (
+from opentrons.protocol_api._core.protocol_api.protocol_context import (
```
Comment (SyntaxColoring) [inline api/src/opentrons/execute.py]: I generally do enjoy underscore-prefixed submodules, but I don't know if this is the best case for it. I'm conflicted.

This is a strange case because:

* **From the perspective of a user-written Python protocol, all of this stuff is unstable and internal.**
    * Users should definitely not be importing or using this stuff.
    * So, in that sense, hiding it under `._core` is good.
* **From the perspective of higher-level Opentrons code, some of this stuff needs to be public.**
    * It's an essential part of running protocols. This `import` line here in `opentrons.execute` is illustrative: this file cannot do its job without importing `ProtocolContextImplementation`. It's a big red flag in the same way as if one class used an underscore-prefixed method from a different class.
    * So, in that sense, hiding it under `._core` is bad, because it reads like this stuff is private to higher-level Opentrons code, and it *can't* be private to higher-level Opentrons code.

This same problem frequently happens at the level of individual methods. For example, `ProtocolContext.equipment_broker` is a public property because our higher-level HTTP API code needs it, but . In that case, we don't underscore-prefix the property, but we're careful to leave it off of docs.opentrons.com.

I think our immediate options are:

* `opentrons.protocol_api._core`, as you currently have it in this PR.
    * For the reasons mentioned above, I softly oppose this. But I'm happy to give it a shot if we want to move on and try something out.
* `opentrons.protocol_api.core`, without the underscore.
    * We'd rely on obscurity to prevent users from finding it, and docstrings to discourage users from using it if they do find it.
    * This is basically our existing strategy for things like `ProtocolContext.equipment_broker`.
* `opentrons.protocol_api_core` (or similar).
    * Basically, we go back to what we had before: we make this a sibling package of `opentrons.protocol_api`.
    * Theoretically, this leaves us open to problems where a user-written protocol can erroneously write `import opentrons.protocol_api_core` and access our internal stuff. But this problem wouldn't be new; it already exists for `opentrons.calibration_storage` and everything else. It's an unavoidable outcome of our public API and our internal stuff sharing the same top-level `opentrons` namespace. So I'm okay with just accepting it.

**Review by SyntaxColoring — COMMENTED** [review]
✅ Smoke-tested with the HTTP API and with `opentrons_simulate`.

File: api/src/opentrons/execute.py (hunk 1 above)
Comment (mcous) [reply api/src/opentrons/execute.py]: After spending more time digging around and [refactoring](https://github.com/Opentrons/opentrons/compare/RCORE-79_move-papiv2-impls...RCORE-78_protocol-engine-papiv2-ff) this code, I think I do, in fact, have a soft preferences for `._core`. **I'm very amenable to `.core`, too, though.**

> - From the perspective of higher-level Opentrons code, some of this stuff needs to be public.

This might be a little pie in the sky from me, but the more time I spend in this code base, the more I think we should continue refactoring the codebase so that this _isn't_ true.

I've got a desire to make it so that **nothing outside of `opentrons.protocol_api` should know about `opentrons.protocol_api._core`**. This PR is a first step, and the next PR will remove a big chunk of imports by hiding `...Implementation` construction behind a public factory method of the `opentrons.protocol_api` module. See [my WIP next PR for more context](https://github.com/Opentrons/opentrons/compare/RCORE-79_move-papiv2-impls...RCORE-78_protocol-engine-papiv2-ff)

**Comment by mcous** [comment]
Something screwy is going on with the commits list of this PR. Closing and recreating

File: api/src/opentrons/execute.py (hunk 1 above)
Comment (SyntaxColoring) [reply api/src/opentrons/execute.py]: Discussed in Slack:

@mcous's above plan makes sense. We're going with `.core` (no underscore) in this PR in case that plan is interrupted and we have to leave the code in an in-between state.

**Review by SyntaxColoring — APPROVED** [review]
Looks good to merge once CI passes. Thanks! :shipit:

File: api/src/opentrons/protocol_api/core/protocol_api/instrument_context.py [hunk 2]
```diff
@@ -230,9 +231,13 @@ def get_available_volume(self) -> float:
     def get_pipette(self) -> PipetteDict:
         """Get the hardware pipette dictionary."""
         sync_hw_api = self._protocol_interface.get_hardware()
-        pipette = sync_hw_api.get_attached_instrument(self._mount)
+        pipette: Optional[PipetteDict] = sync_hw_api.get_attached_instrument(
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_api/core/protocol_api/instrument_context.py:234]: why is this optional? I looked into the method and its returning a dict but not optional? but I am also confused with the HardwareApi implementation so I could be missing something?

File: api/src/opentrons/protocol_api/core/protocol_api/protocol_context.py [hunk 3]
```diff
@@ -99,7 +95,7 @@ def get_hardware(self) -> SyncHardwareAPI:
 
     def is_simulating(self) -> bool:
         """Returns true if hardware is being simulated."""
-        return self._sync_hardware.is_simulator
+        return self._sync_hardware.is_simulator  # type: ignore[no-any-return]
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_api/core/protocol_api/protocol_context.py:98]: why do we need this type ignore?

**Review by TamarZanzouri — APPROVED** [review]
just a few questions but overall looks great! thank you for doing this! I think it will save us headaches in the future

File: api/src/opentrons/protocol_api/core/protocol_api/protocol_context.py (hunk 3 above)
Comment (mcous) [reply api/src/opentrons/protocol_api/core/protocol_api/protocol_context.py:98]: I increased the type-checking strictness of this code, so this is an existing issue that the type-checker already knew about but just wasn't complaining about.

The root of the issue is that the `SynchronousAdapter` erases type signatures. We've got RSS-52 on the board to address that

File: api/src/opentrons/protocol_api/core/protocol_api/instrument_context.py (hunk 2 above)
Comment (mcous) [reply api/src/opentrons/protocol_api/core/protocol_api/instrument_context.py:234]: Oh yikes. I saw that later `is None`, and assumed it could be `None`, but you're right, the hardware API will not return `None`.

This looks like an existing bug, I will look into addressing in a follow up PR
